backends/custom_cpu/tests/unittests/custom_relu_op.cc

```diff
@@ -0,0 +1,202 @@
+  return {out};
+}
+
+std::vector<paddle::Tensor> relu_custom_backward(
+    const paddle::Tensor& x,
+    const paddle::Tensor& out,
+    const paddle::Tensor& grad_out) {
+  CHECK_CUSTOM_INPUT(x);
+  CHECK_CUSTOM_INPUT(out);
+  // CHECK_CUSTOM_INPUT(grad_out);
+  auto grad_x = paddle::empty_like(x, x.dtype(), x.place());
+  auto ones = paddle::experimental::full_like(x, 1.0, x.dtype(), x.place());
+  auto zeros = paddle::experimental::full_like(x, 0.0, x.dtype(), x.place());
+  auto condition = paddle::experimental::greater_than(x, zeros);
+
+  grad_x = paddle::multiply(grad_out, paddle::where(condition, ones, zeros));
+
+  return {grad_x};
+}
+
+std::vector<paddle::Tensor> relu_custom_double_backward(
+    const paddle::Tensor& out, const paddle::Tensor& ddx) {
+  CHECK_CUSTOM_INPUT(out);
+  auto ddout = paddle::empty(out.shape(), out.dtype(), out.place());
+  auto ones =
+      paddle::experimental::full_like(out, 1.0, out.dtype(), out.place());
+  auto zeros =
+      paddle::experimental::full_like(out, 0.0, out.dtype(), out.place());
+  auto condition = paddle::experimental::greater_than(out, zeros);
+
+  ddout = paddle::multiply(ddx, paddle::where(condition, ones, zeros));
+  std::cout << "DEBUG custom ddout shape" << ddout.numel() << std::endl;
+
+  return {ddout};
+}
+
+std::vector<paddle::Tensor> ReluForward(const paddle::Tensor& x) {
+  if (x.is_cpu()) {
+    return relu_cpu_forward(x);
+  } else if (x.is_custom_device()) {
+    std::cout << "Debug info: run custom relu cpu forward." << std::endl;
+    return relu_custom_forward(x);
+  } else {
+    PD_THROW("Not implemented.");
+  }
+}
+
+std::vector<paddle::Tensor> ReluBackward(const paddle::Tensor& x,
+                                         const paddle::Tensor& out,
+                                         const paddle::Tensor& grad_out) {
+  if (x.is_cpu()) {
+    return relu_cpu_backward(x, out, grad_out);
+  } else if (x.is_custom_device()) {
+    std::cout << "Debug info: run custom relu cpu backward." << std::endl;
+    return relu_custom_backward(x, out, grad_out);
+  } else {
+    PD_THROW("Not implemented.");
+  }
+}
+
+std::vector<paddle::Tensor> ReluDoubleBackward(const paddle::Tensor& out,
+                                               const paddle::Tensor& ddx) {
+  if (out.place() == paddle::PlaceType::kCPU) {
+    std::cout << "Debug info: run cpu double backward." << std::endl;
+    return relu_cpu_double_backward(out, ddx);
+  } else if (out.place().GetType() == phi::AllocationType::CUSTOM) {
+    std::cout << "Debug info: run custom double backward." << std::endl;
+    return relu_custom_double_backward(out, ddx);
+  } else {
+    PD_THROW("Not implemented.");
+  }
+}
+
+std::vector<std::vector<int64_t>> ReluDoubleBackwardInferShape(
+    const std::vector<int64_t>& out_shape,
+    const std::vector<int64_t>& ddx_shape) {
+  return {out_shape};
+}
+
+PD_BUILD_OP(custom_relu)
+    .Inputs({"X"})
+    .Outputs({"Out"})
+    .SetKernelFn(PD_KERNEL(ReluForward));
```

Comment:
SetKernelFn是为哪个设备注册kernel？

Reply:
`SetKernelFn` 不区分设备，会为所有的设备都注册 kernel，开启 paddle `WITH_CUSTOM_DEVICE` 编译选项后会自动注册自定义设备

`SetKernelFn` does not distinguish devices, it will register kernels for all available devices. Custom device kernel will be registered automatically when opens `WITH_CUSTOM_DEVICE` flag

Reply:
不区分设备的话，是否支持多个插件注册同名的op？

Reply:
支持多个不同设备的 kernel 实现，实际上不用注册多个 kernel。以本单测为例，对于设备的区分封装在 ReluForward 函数中，ReluForward 函数里面根据不同的 place 走不同的分支。

We support kernels with different devices. Actually, users do not need to register multiple kernels. For example, the function `ReluForward` distinguishes different devices, `ReluForward` will visit different logic branches according to the tensor place.